### PR TITLE
[TK-04399] Validation Package ground work

### PR DESCRIPTION
**Pull request by freesig**
Adds the ground work for validation packages.
The can be set in the `EntrtyDef` to either:
- Element: Just the element
- SubChain: All elements in the chain with the same entry type
- FullChain: The entire chain
Only `Element` is implemented for this PR as the other two require additional network calls to be added to get the packages.

other notable change is validation callbacks now take `ValidateData` instead of an `Element` so we can pass in the `ValidationPackage` (If it's not just the element) and the `Element`.

## Usage
Field: `required_validation_package`.
Options: `element, sub_chain or full`.
```rust
#[hdk_entry(id = "post", required_validations = 5, required_validation_package = "element")]
struct Post(String)
```

### CODE REVIEW

File: crates/hdk_derive/src/lib.rs [hunk 1]
```diff
@@ -40,6 +43,26 @@ impl Parse for EntryDef {
                         }
                         _ => unreachable!(),
                     },
+                    "required_validation_package" => {
```
Comment (maackle) [inline crates/hdk_derive/src/lib.rs:46]: Maybe just "validation_package" for the hdk specifier?

```suggestion
                    "validation_package" => {
```

**Review by maackle — APPROVED** [review]
Looks good, and I know it's groundwork so I won't comment on the fact that ValidationPackage is empty :)

I wonder if this is a good place to include some of those other layers of validation we have talked about. Like, `FullChain` validation implies that the chain is overall valid, e.g. no fork has been detected. Maybe there might also be a `ElementHistory` package which presupposes that the entire redirect history is valid. Just thoughts for the future.

File: crates/hdk_derive/src/lib.rs (hunk 1 above)
Comment (freesig) [reply crates/hdk_derive/src/lib.rs:46]: works for me

**Comment by freesig** [comment]
> Looks good, and I know it's groundwork so I won't comment on the fact that ValidationPackage is empty :)
> 
> I wonder if this is a good place to include some of those other layers of validation we have talked about. Like, `FullChain` validation implies that the chain is overall valid, e.g. no fork has been detected. Maybe there might also be a `ElementHistory` package which presupposes that the entire redirect history is valid. Just thoughts for the future.

I'm not so sure about this actually.
I think there is nothing about a validation package that says the chain is valid.
If they are online then you are hoping to get it from the author. So if they wanted to send you an invalid chain they easily could.
The only way to know if the chain is valid is to ask an Agent Activity authority but they are the worst person to ask because they don't have the entries so you then have to do a get for each entry.

I think if you have an app that requires the chain be valid _at the time_ that you are validating X where X requires full chain then you need to validate the chain in the app validation callback.

**Pull request merged**